kong/llm/init.lua

```diff
@@ -252,6 +381,159 @@ function _M.is_compatible(request, route_type)
+  ngx.header["Via"] = meta._SERVER_TOKENS
+
+  for k, v in pairs(res.headers) do
+    if not streaming_skip_headers[lower(k)] then
+      ngx.header[k] = v
+    end
+  end
+
+  -- server-sent events should ALWAYS be chunk encoded.
+  -- if they aren't then... we just won't support them.
+  repeat
+    -- receive next chunk
+    local buffer, err = reader(buffer_size)
+    if err then
+        ngx.log(ngx.ERR, err)
+        break
+    end
+
+    if buffer then
+      -- we need to rip each message from this chunk
+      events = {}
+      for s in buffer:gmatch("[^\r\n]+") do
+        table.insert(events, s)
+      end
+
+      local metadata
+
+      -- then parse each into the standard inference format
+      for i, event in ipairs(events) do
+        local event_t
+        local token_t
+
+        -- some LLMs do a final reply with token counts, and such
+        -- so we will stash them if supported
+        local formatted, err, this_metadata = self.driver.from_format(event, self.conf.model, "stream/" .. self.conf.route_type)
+        if err then
+          return internal_server_error(err)
+        end
+
+        metadata = this_metadata or metadata
+
+        -- handle event telemetry
+        if self.conf.logging.log_statistics then
+
+          if not ai_shared.streaming_has_token_counts[self.conf.model.provider] then
+            event_t = cjson.decode(formatted)
+            token_t = get_token_text(event_t)
+
+            -- incredibly loose estimate based on https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them
+            -- but this is all we can do until OpenAI fixes this...
+            --
+            -- essentially, every 4 characters is a token, with minimum of 1 per event
+            telemetry_fake_table.usage.completion_tokens =
+                telemetry_fake_table.usage.completion_tokens + math.ceil(#string.gsub(token_t, "^%s*(.-)%s*$", "%1") / 4)
+
+          elseif metadata then
+            telemetry_fake_table.usage.completion_tokens = metadata.completion_tokens
+            telemetry_fake_table.usage.prompt_tokens = metadata.prompt_tokens
+          end
+
+        end
+
+        -- then stream to the client
+        if formatted then  -- only stream relevant frames back to the user
+          if self.conf.logging.log_payloads then
+            -- append the "choice" to the buffer, for logging later. this actually works!
+            if not event_t then
+              event_t = cjson.decode(formatted)
+            end
+            
+            if not token_t then
+              token_t = get_token_text(event_t)
+            end
+
+            if err then
+              return internal_server_error("something wrong with decoding a specific token")
+            end
+
+            telemetry_fake_table.response:put(token_t)
+          end
+
+          -- construct, transmit, and flush the frame
+          ngx.print(fmt("data: %s\n\n", formatted))
+          ngx.flush()
```

Comment:
You _may_ want to invoke this as `ngx.flush(true)` if I am reading [the docs](https://github.com/openresty/lua-nginx-module?tab=readme-ov-file#ngxflush) correctly:

> When `ngx.flush(true)` is called immediately after [ngx.print](https://github.com/openresty/lua-nginx-module?tab=readme-ov-file#ngxprint) or [ngx.say](https://github.com/openresty/lua-nginx-module?tab=readme-ov-file#ngxsay), it causes the latter functions to run in synchronous mode. This can be particularly useful for streaming output.

Reply:
Fixed in ba247aa5a7624b835e68f867efb70f0e0e55dfb3